Chardata docs (#7194)
SciTools/iris: docs/src/user_manual/explanation/netcdf_io.rst

    (70, 37, 49)

Character and String datatypes

> ------------------------------

Reviewer: I think an introduction might be warranted. Something like:

In NetCDF, text can be represented in 2 ways:
- `char`: the number of characters in the text is fixed
- `string`: the number of characters in the text can vary

Then text can have different *encodings*, which affects the number of *bytes* used to store a character. For example:
- `ASCII` and `UTF-32`: the number of bytes are fixed
- `UTF-8` and `UTF-16`: the number of bytes can vary

This section discusses how Iris handles the different cases and how the standards and interpretations have changed.

Summary
^^^^^^^

> *   Iris currently *only* fully supports fixed-width 'char' type data in netCDF variables

Reviewer: Are the single quotes intended to be backticks? It doesn't look like the rest of the document uses single quotes for this.

Author: OK I was, in theory, intending to use this style _specifically_ for the netCDF type names.
But I think I'll follow your https://github.com/SciTools/iris/pull/7194#discussion_r3552503232 and convert them all to double backticks aka 'code format'. 
https://github.com/SciTools/iris/pull/7194/commits/ff827f1a497f4b6cde2468d064d069d11edc703b

>     * the 'string' type (variable-width unicode strings) will be added in a future release

Reviewer: A link to an issue to track this might be useful

Author: https://github.com/SciTools/iris/pull/7194/commits/82e3588aa273f691e04bb9d05c9956d408ae1b0a

*   'char' variable data is represented as numpy string arrays in Iris objects, such as
    cubes and coordinates.

>     *   the numpy dtype is of the type "U<xx>", where <xx> is a character width.

Reviewer: This is referred to as "string width" later on. I think that term is more correct?

Author: https://github.com/SciTools/iris/pull/7194/commits/dfe771482b08e1cb9802f8ffbfae2c18dca8a12a

>     *   the character width relates to a string *dimension* of the netCDF variable,

Reviewer: I'm not following what this bullet point means.

Author: https://github.com/SciTools/iris/pull/7194/commits/dfe771482b08e1cb9802f8ffbfae2c18dca8a12a

        which is not in the dimensions of the Iris object or its data array.
    *   the dtype 'width' controls the length of string dimensions created when saving

*   Iris also uses a variable ``_Encoding`` attribute to enable storage of non-ascii
    characters in 'char' type arrays.

    *   it appears as a regular attribute of the Iris object
    *   it is not needed for ascii-only data
    *   it is not needed to *read* 'utf-8' encoded data correctly
    *   it **is** required to *save* any non-ascii characters

The following describes the nature of character and string data handling in :

> netCDF itself; the CF conventions; the netCDF4 Python module and the Iris implementation.

Reviewer: NetCDF is inconsistently capitalised

Author: https://github.com/SciTools/iris/pull/7194/commits/da24fe9fbe6f94f28567aea184a4017d8200bf7c

Note that the Python module is called ``netCDF4``.

storage characteristics of character data are relevant:

*   The **names** of file components (variables, dimensions, and attributes) are
    natively unicode-capable strings of arbitrary (variable) length.

*   **Attributes** with string content likewise *appear* to be natively unicode.  However,
    the actual datatype of the attribute may vary, being either 'char' or 'string'.

*   The **content of variables** can be either 'char' or 'string'.

    *   'string' type variables contain a variable-length unicode string at each array element.

    *   'char' type variables contain one-byte characters, and generally have a fixed-length
        "string dimension".  If they contain *only* ascii character values, this is
        uncomplicated, but they may also be used to contain non-ascii data (i.e.
        including unicode characters).  There is no universally defined agreement for
        how to indicate that bytes are encoded non-ascii data, but many older datasets
        have used a variable attribute ``_Encoding`` indicating the encoding name.

The NetCDF documentation also mentions that an ``_Encoding`` attribute may be used to

> represent non-ascii strings.  However this is described as "reserved for future use",

Reviewer: Citation would be useful

Author: https://github.com/SciTools/iris/pull/7194/commits/82e3588aa273f691e04bb9d05c9956d408ae1b0a

and its valid values and effects are not explicitly defined.

However, it is also notable that the standard ``ncgen`` and ``ncdump`` tools *do*
correctly interpret an ``_Encoding`` attribute in most cases, despite this not being an
"official" solution.

In the netCDF CF Conventions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The `CF Conventions <https://https://cfconventions.org/>`_ define a subset of

Reviewer: This link is broken

Author: https://github.com/SciTools/iris/pull/7194/commits/82e3588aa273f691e04bb9d05c9956d408ae1b0a

store).

Although the variable-length 'string' data is now supported in CF, the use of
fixed-width 'char' arrays is obviously more efficient for storage and access, and it is
still the most common approach in practice.

In the netCDF4 Python module
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

*   attributes with string content always appear as Python 'str' (i.e. unicode strings).
    It is not possible to distinguish or control the 'char' and 'string' type in the file,
    this is hidden from the user by the Python implementation.

*   variable data of type 'string' is presented (read and written) as numpy arrays of dtype
    "U<xx>", where <xx> is a (maximum) string length.

*   variable data of type 'char' is presented (read and written) as numpy arrays of
    dtype "S1" -- that is, an array of length-1 Python "bytes" objects.

    .. note::

        The netCDF4 package can also automatically translate this to string arrays of
        dtype "U<xx>", if the variable has an ``_Encoding`` attribute.  **However,**
        Iris turns this feature *off*, in order to implement its own wider-ranging
        encoding support (described below).

> In Iris

Reviewer: I think the Iris part of this section should be the first under summary, since I imagine users are most interested in that. The other sections are more to provide context for decisions made in Iris

^^^^^^^
.. note::

    In Iris, **the 'string' data type is not supported at present**, though this is

>     planned for future releases.  See the following section `Variable-length datatypes`_

Reviewer: Same as previous comment for link to issue

Author: https://github.com/SciTools/iris/pull/7194/commits/82e3588aa273f691e04bb9d05c9956d408ae1b0a

    for an interim solution enabling you at least to *load* variable-length string data.

Iris stores string data in arrays of dtype "U<xx>", where <xx> is a maximum character
width.  However, this data is currently **only** read and written in netCDF files as
'char' type variables (i.e. byte arrays).

Iris supports a specific set of valid encodings for non-ascii data :

> "ascii", "utf8", "utf16" and "utf32".  These (or aliases) will appear in the

Reviewer: This reads like ascii is non-ascii data

Author: https://github.com/SciTools/iris/pull/7194/commits/dfe771482b08e1cb9802f8ffbfae2c18dca8a12a

``_Encoding`` attribute of a file variable, and likewise as an attribute of the
corresponding Iris component object (e.g. cube or coordinate).

When loading
~~~~~~~~~~~~
If there is a valid ``_Encoding`` attribute this is used to decode the
data, otherwise a default encoding of "utf8" is applied:  This works transparently when
only ascii characters are present, and also allows the ``_Encoding`` attribute to be
omitted as long as utf8 was used.  An invalid or unsupported encoding name will be
ignored, with a warning, but the attribute will still be added to the Iris component
object.

When saving
~~~~~~~~~~~
Any string data with only ascii characters does not require an ``_Encoding`` attribute.
However if there are any non-ascii characters, and no ``_Encoding``

> attribute, then an error will be raised.  An invalid or unsupported encoding name will

Reviewer: Might be useful to have an example of how to set the `_Encoding` attribute

Author: https://github.com/SciTools/iris/pull/7194/commits/dfe771482b08e1cb9802f8ffbfae2c18dca8a12a

be ignored, with a warning, but the attribute will still be stored to the file.

So effectively, the **default encoding is 'utf8' for load and 'ascii' for save**.

String width dimensions
~~~~~~~~~~~~~~~~~~~~~~~
For each valid encoding there is a definite relation between the string dimension length
in the file (actually, the number of *bytes*), and the maximum character length, aka
string width, in the array dtype : i.e. the "<xx>" in the "U<xx>" dtype.

The **lengths of string dimensions created on write** are calculated as follows:

*   ascii : n-bytes = n-characters

> *   utf8 : n-bytes = n-characters

Reviewer: I think it's worth emphasising that utf8 and utf16 examples are minimums in the list as well, rather than just in a later sentence

Author: https://github.com/SciTools/iris/pull/7194/commits/dfe771482b08e1cb9802f8ffbfae2c18dca8a12a

    (70, 37, 49)

Character and String datatypes
------------------------------
Overview
^^^^^^^^
Text appears in NetCDF in a variety of ways, but the key aspect to be explained here is
the storage of bulk text data in *variables*.

In Iris, all string data appears in Iris' data objects as arrays of numpy dtype "U<xx>",

> where <xx> is a maximum character width.  When reading and writing NetCDF files, only

Reviewer: ```suggestion
where <xx> is a maximum string width.  When reading and writing NetCDF files, only
```

the NetCDF ``char`` datatype is currently supported.

String support is fairly simple when strings contain only ASCII characters.
When strings may include non-ascii characters, this requires a specific encoding to be
adopted when translating to and from bytes, and rules for determining what the encoding
is or was.

> Background

Reviewer: I think the Background can now be moved above "String data in NetCDF", "String data in the NetCDF CF Conventions" and "String data in the netCDF4 Python module" and make them subheadings

Author: OK that's nice + a simpler way of doing it than creating a separate new page.

~~~~~~~~~~
The relevant supporting code libraries and standards provide various facilities for
translating between bytes and Python/numpy strings, but not all possibilities are
supported.  The facilities and conventions for this have changed over time, and
obsolete methods persist in archive datasets, which must therefore be taken into
consideration.

This documentation explains how Iris handles the different cases, and also relevant
aspects of the supporting projects, which in practice affect its design.  These are:

*   the NetCDF file format;
*   the CF conventions; and
*   the ``netCDF4`` Python module.

String data in Iris
^^^^^^^^^^^^^^^^^^^
Iris stores string data in arrays of numpy dtype "U<xx>", where <xx> is a maximum

> character width.  This data is currently **only** read and written to NetCDF files as

Reviewer: ```suggestion
string width.  This data is currently **only** read and written to NetCDF files as
```

characters present **and can exceed the numbers given above**.

String widths in Saving
#######################
If any string in an actual data array encodes to *more* bytes than the above-calculated
string dimension, when written, then Iris will raise an
:class:`iris.exceptions.TranslationError`.  In this case, the user should **explicitly
specify** a longer string dimension, by converting the data to a longer "U<xx>" dtype :
for example, ``cube.data = cube.core_data().astype("U20")``.

.. warning::

    When processing string arrays, Numpy does not routinely preserve the "<xx>" width part
    of "U<xx>" type data : instead, some operations will reduce it to the maximum width
    occurring.  So in these cases also, it may be necessary to explicitly re-assert the
    desired "string width" before saving -- use ``.astype()``,  as above.

String widths in Loading
########################
On reading, the returned data has a '"U<xx>"' dtype of which the <xx> string width is
determined by **inverting the above relations**.

The actual maximum number of characters in the data will always be less than or equal to

> this, since the maximum possible character width is for all-ascii data.

Reviewer: I assume you meant to say 
```suggestion
this, since the maximum possible string width is 1 byte for all-ascii data.
```

Author: Well, not exactly..
I meant that the longest string you can get is when the bytes are full with all ascii characters.
I'll try + rephrase this more clearly...

The width created by reading will always round-trip correctly, i.e. the dimension
length is unchanged if data is read and then written back.

> Summary of Iris string handling

Reviewer: I think this summary can be removed now and once you have more concrete examples (mentioned in private), something like "Iris string handling in practice" would be a good replacement.

**any** variables.
However, *historically*, CF had more limited support, and also "unofficial conventions"
have been used for string data encoded as bytes, which may be encountered
in older datasets, as follows ...

Prior to v1.8
~~~~~~~~~~~~~
CF required to use ``char`` type only, and provided
**no official means** of representing non-ascii data.

Since v1.8
~~~~~~~~~~
CF has allowed the use of ``string`` data in all variables.
However, up to v1.12 there was still no official way of encoding non-ascii data in
``char`` arrays.

Since v1.12
~~~~~~~~~~~
CF now mandates a *default* assumption of utf-8 encoding to store
non-ascii data in ``char`` form.  It does also note that some data in the past has used an
``_Encoding`` attribute -- though this was never an official CF usage.

Characteristics of CF string storage
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Where strings are stored as ``char`` datatype, which is the more common traditional approach,
the array must have a "string dimension", which is a normal file dimension.  Thus, these
strings always have a *fixed byte width*.  However, that is not the same as a fixed

> *character* width, since in most encodings non-ascii characters require more bytes to

Reviewer: ```suggestion
*string* width, since in most encodings non-ascii characters require more bytes to
```

still the most common approach in practice.

String data in the netCDF4 Python module
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

*   attributes with string content always appear as Python 'str' (i.e. unicode strings).
    It is not possible to distinguish or control the ``char`` and ``string`` datatype in the file,
    this is hidden from the user by the Python implementation.

*   variable data of type ``string`` is presented (read and written) as numpy arrays of dtype
    "U<xx>", where <xx> is a (maximum) string width.

*   variable data of type ``char`` is presented (read and written) as numpy arrays of
    dtype "S1" -- that is, each element is a single byte, which reads as a length-1
    Python "bytes" object.

    .. note::

        The netCDF4 package can also automatically translate byte arrays into string
        arrays of dtype "U<xx>" on load, if the variable has an ``_Encoding`` attribute.
        See in netCDF4 python documentation :
        `Dealing with strings <https://unidata.github.io/netcdf4-python/#dealing-with-strings>`_.
        **However,** Iris turns this feature *off*, in order to implement its own

>         wider-ranging encoding support (described below).

Reviewer: no longer described below

    (70, 37, 49)

Character and String datatypes
------------------------------
Text can be present  in NetCDF in a variety of ways

> (see : `String data in NetCDF <cf_strings>`_ for details).

Reviewer: Link is broken. I think it should be 

```suggestion
(see : `String data in NetCDF <cf-strings>`_ for details).
```

The main aspect to be explained here is the storage of bulk text data in *variables*.

String data in Iris
^^^^^^^^^^^^^^^^^^^
Iris objects can store strings in their data arrays, such as a cube ``.data`` or
coordinate ``.points``.

These are always stored as arrays of numpy dtype "U<xx>", where <xx> is a maximum

> string width (which could be either `real or lazy <real_and_lazy_data>`_).

Reviewer: Link is broken. I think it should be 
```suggestion
string width (which could be either `real or lazy <real-and-lazy_data>`_).
```
